### PR TITLE
Add Decred(DCR)

### PR DESCRIPTION
**Pull request by cipherzzz**
Hello, Thanks for providing this great resource. I am looking to add Decred to electrumX and would love your feedback on getting this change set integrated.

### CODE REVIEW

File: lib/coins.py [hunk 1]
```diff
@@ -43,6 +43,7 @@
 from server.block_processor import BlockProcessor
 import server.daemon as daemon
 from server.session import ElectrumX, DashElectrumX
+from functools import partial
```
Comment (kyuupichan) [inline lib/coins.py]: Python stdlib imports go in the first block of imports, as per generally accepted coding style

**Comment by kyuupichan** [comment]
Merged, thanks!

**Comment by cipherzzz** [comment]
awesome, thanks!

**Comment by erasmospunk** [comment]
@cipherzzz have you managed to fully sync the blockchain? I haven't tested this but this is my [code](https://github.com/erasmospunk/electrumx/commit/878a7b0344b134ffbad0f2ea29249b9762e7f4f1) and I know that it doesn't work (you left `ALLOW_ADVANCING_ERRORS = True` but don't have the relative code in `block_processor.py`).

Also why on earth you removed my commit and in effect took the credit?

@kyuupichan ?

**Comment by cipherzzz** [comment]
I don't know what commit of yours I removed, but it was certainly through
ignorance and not trying to take credit.
I apologize for that - I'm just trying to get decred up...  I do not have a
ton of open source experience and am learning as I go. Once again,
apologies if I stepped on your toes

**Comment by kyuupichan** [comment]
So does it sync?  Sorry @erasmospunk I assume if someone's submitting a coin implementation that it works; I can't really check them all myself.

**Comment by cipherzzz** [comment]
It seemed to be syncing for me...

**Comment by cipherzzz** [comment]
@erasmospunk - I gave you props in this article I wrote a while ago on working through syncing decred. If it's not syncing for you, maybe take a look at this: https://medium.com/@cipherz/developing-on-electrumx-d1342ab3c8b5

**Comment by erasmospunk** [comment]
@cipherzzz understood, next time try `git rebase` and in general preserve the commit history if you are improving the work of others.

@kyuupichan testing everything is infeasible, I'm not sure how to solve this problem. Just tried it and it does not fully sync:
```
2018-03-11 17:15:07,721 INFO    [58] RPC 127.0.0.1:54344, 1 total                                                                    [util.py:57]
2018-03-11 17:15:07,854 INFO    our height: 80,009 daemon: 220,091 UTXOs 199MB hist 45MB                                             [block_processor.py:485]
2018-03-11 17:15:52,791 ERROR   Traceback (most recent call last):
  File "/coinomi/electrumx/server/controller.py", line 176, in on_future_done
    future.result()
  File "/coinomi/electrumx/server/block_processor.py", line 206, in main_loop
    await task()
  File "/coinomi/electrumx/server/block_processor.py", line 250, in check_and_advance_blocks
    await self.controller.run_in_executor(self.advance_blocks, blocks)
  File "/coinomi/electrumx/server/controller.py", line 156, in run_in_executor
    return await self.loop.run_in_executor(None, func, *args)
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/coinomi/electrumx/server/block_processor.py", line 502, in advance_blocks
    undo_info = self.advance_txs(block.transactions)
  File "/coinomi/electrumx/server/block_processor.py", line 543, in advance_txs
    cache_value = spend_utxo(txin.prev_hash, txin.prev_idx)
  File "/coinomi/electrumx/server/block_processor.py", line 728, in spend_utxo
    .format(hash_to_str(tx_hash), tx_idx))
server.block_processor.ChainError: UTXO 38c81dd1cff07f3fb4d628b77fc234dd65bc812309ba9ce486ca1a3d852a063d / 2 not found in "h" table
 [util.py:63]
```

The problem is that Decred have a special "voting" block that invalidates the transactions from the previous block. This means that a transaction spent TXO `38c81dd1cff07f3fb4d6... / 2`, got invalidated and when another transaction spent the same TXO, but it wasn't present in the "h" table in ElectrumX.

**Comment by kyuupichan** [comment]
@cipherzzz I assume the failure @erasmospunk gets is real - so what is the point of this if it doesn't really work?

**Comment by cipherzzz** [comment]
@kyuupichan @erasmospunk 
I was able to sync the testnet so I assumed it was ok
**Electrum**
```
2018-03-11 17:58:18,105 INFO    flush #202 took 0.0s.  Height 257,313 txs: 3,216,092                                                 [block_processor.py:395]
2018-03-11 17:58:18,105 INFO    processed 1 block in 0.0s 
```

**DCRD**
```
2018-03-11 17:58:16.900 [INF] BMGR: Processed 1 block in the last 23.57s (3 transactions, 7 tickets, 5 votes, 0 revocations, height 257313, 2018-03-11 17:58:15 -0400 EDT)
```
However, I do get the failure at height 80k on mainnet as well
```
server.block_processor.ChainError: UTXO 38c81dd1cff07f3fb4d628b77fc234dd65bc812309ba9ce486ca1a3d852a063d / 2 not found in "h" table
 [util.py:63]
```

I can ask the Decred team tomorrow if they have any ideas, but obviously it is not ready for prime time.

**Comment by kyuupichan** [comment]
@cipherzzz OK, if you would like it to stay in please get it working, otherwise I will remove in the next release.  There's little point releasing something that doesn't sync.

**Pull request merged**